### DISCUSSION TITLE
Modify parameter to be variable parameter for sensitivity analysis with the ospsuite R package

### DISCUSSION THREAD

**Discussion by moPellowe** [Open-Systems-Pharmacology/Forum#1754] [Questions & Problems] [answered]
I have noticed that some parameter paths are not available for sensitivity analysis in R. I can look at the list of variable parameters using the function `potentialVariableParameterPathsFor()`, but how do I add parameters of interest to this list? I have tried going into the pkml and modifying the flag for the parameters I want so that `ReadOnly = FALSE` and `CanBeVaried == TRUE`, i.e. `flag %% 8 == 2`. However, this only works for some of the parameters that I want. What exactly are the criteria for a parameter to be valid for sensitivity analysis in R?

**Comment by Yuri05** [comment, marked as answer]
The flag is a bit-field where each bit represents a specific state of the parameter in PK-Sim/MoBi. 
If the n-th bit is set to 1: the flag is enabled, otherwise disabled.

The states (bits) are defined as following
```
CanBeVaried = 2,
ReadOnly = 4,
Visible = 8,
MinIsAllowed = 16,
MaxIsAllowed = 32,
CanBeVariedInPopulation = 64,
IsChangedByCreateIndividual = 128,
```

Example:
```
<Parameter id="xxx" ... >
    <Info flag="52" ... />
</Parameter>
```
Here: 52 = 32+16+4 meaning that the parameter is defined as:
```
CanBeVaried = FALSE
ReadOnly = TRUE
Visible = FALSE
MinIsAllowed = TRUE
MaxIsAllowed = TRUE
CanBeVariedInPopulation = FALSE
IsChangedByCreateIndividual = FALSE
```

The visible flag is used by PK-Sim only, but if one day you send a simulation from MoBi back to PK-Sim parameter will disappear in the UI, so we should set it as visible as well.

So we need to make 3 changes
a)	Switch the CanBeVaried state from FALSE to TRUE ➡️  Add 2
b)	Switch the ReadOnly state from TRUE to False ➡️  Substract 4
c)	Switch the Visible state from FALSE to TRUE ➡️ Add 8

So in total: +2 -4 +8 ➡️ Add 6 to the original flag to make you parameter appear everywhere (in the PI etc.)

So the modified parameter chunk in the pkml file will be:

```
<Parameter id="xxx" ... >
    <Info flag="58" ... />
</Parameter>
```

If after that the parameter still does not appear in the sensitivity analysis, it might have other reasons, e.g.
* parameter value = 0
* parameter is defined by a table 
* etc.

**Reply by moPellowe** [reply]
Thank you for the detailed explanation.

Can you help me understand why "Ki" is not in the list of parameters that can be varied? I've attached a test pkml where there is an inhibition process. The flag for Ki is already 106 (64+32+8+2), so I believe that all of the requirements mentioned above are satisfied.

![image](https://github.com/Open-Systems-Pharmacology/Forum/assets/57540473/fccf5814-9db1-44f2-b2c8-1a24b696f397)

Yet, it is not in the list of potential variable parameter paths.

![image](https://github.com/Open-Systems-Pharmacology/Forum/assets/57540473/7424bcd4-f6e8-491d-aa2c-a3fe1c2ec1f1)

![image](https://github.com/Open-Systems-Pharmacology/Forum/assets/57540473/77bb21a0-9356-4570-8499-196c6ecf7a52)




[temp.zip](https://github.com/Open-Systems-Pharmacology/Forum/files/15069821/temp.zip)

**Reply by Yuri05** [reply]
After looking at your simulation: I don't see any reason why the sensitivity calculation for Ki should not be available - and in fact you can calculate its sensitivity by explicitly adding the parameter path, e.g.

```
> sim <- loadSimulation("test.pkml")
> outputPath="Organism|PeripheralVenousBlood|Alfentanil|Plasma (Peripheral Venous Blood)"
> sim$outputSelections$addQuantity(getQuantity("Organism|PeripheralVenousBlood|Alfentanil|Plasma (Peripheral Venous Blood)", sim))
> sa=SensitivityAnalysis$new(simulation = sim, parameterPaths = c("Alfentanil|CYP3A4-test|Ki"))
> saResult <- runSensitivityAnalysis(sa)
> print(saResult)
SensitivityAnalysisResults: 
   Number of calculated sensitivities: 11 
   Available PK parameters: C_max t_max C_tEnd AUC_tEnd AUC_inf MRT Thalf FractionAucLastToInf CL Vss Vd 
> saResult$pkParameterSensitivityValueFor(pkParameterName = "AUC_tEnd", outputPath = outputPath, parameterPath = "Alfentanil|CYP3A4-test|Ki")
[1] 0
```

I cannot explain why the parameter does not show in the list returned by `potentialVariableParameterPathsFor()` and would consider it as a bug.

**Reply by moPellowe** [reply]
Okay, yes. I have tried to run the sensitivity analysis now and it works despite the parameter not showing up in the list returned by `potentialVariableParameterPathsFor()`. Thanks, @Yuri05 .

**Reply by msevestre** [reply]
@moPellowe Looking at this. It looks that the parameter is DEFINED in the simulation however is not used anywhere. Is this correct? (we analyze the simulation and remove all parameters that are just defined but not referenced). that would be the case if the process is not instantiated for instance. can you confirm?

**Reply by msevestre** [reply]
and indeed. the parameter is not referenced in any equations. This is by design so that you only get the list of parameters relevant for your sensitivity. SA for this parameter should be 0

**Reply by moPellowe** [reply]
@msevestre , yes, you are correct. I have now tested the case where the parameter is used in the simulation, and it does appear in the output of `potentialVariableParameterPathsFor()` after the modification to the flag. Thank you for clarifying that.
